### ISSUE TITLE
Pull request 449 (associated with issue 446) breaks applications in the field and needs to be revisited

### ISSUE THREAD

**Issue by KRyden** [opengeospatial/geopackage#498]
Pull request #449's description is vague but is associated with a larger discussion in #446.  There are several issues with this pull request - which has already been integrated into the working Geopackage document.

Issue#1 - the pull request removes the AUTOINCREMENT.  AUTOINCREMENT is used in fielded software to allow SQLite to manage the unique row ID.  The following code already in use in the field will no longer work without AUTOINCREMENT:

CREATE TABLE example1 (
  id   INTEGER PRIMARY KEY,
  tag  TEXT NOT NULL,
  geom POINT);

And then insert like so:

INSERT INTO example1 (tag,geom) 
VALUES (‘Row 1’,ST_GeomFromText(‘POINT (1 1)’,4326));

In this example, ID is autogenerated by SQLite - without this, the application would have to manage the rowid.

Issue#2: AUTOINCREMENT is still referenced in several tables - was the intent to remove it everywhere (which would be bad as discussed above)?

Issue#3: Requirement 150 probably does not meet modular specification requirements:

Requirement 150
A feature table or view SHALL have a column that uniquely identifies the row. For a feature table, the column SHOULD be an integer primary key. If there is no primary key column, the first column SHALL be of type INTEGER and SHALL contain unique values for each row.

The "SHOULD" would prevent Requirement 150 from being a testable requirement, and really doesn't add anything.  this needs to be reviewed.

Issue#4: it's not clear that quorum was present for the adoption of this change.  The wiki page shows 11 voting members, but only five present - 6 would be needed for quorum (50%, but odd number so +1). https://portal.opengeospatial.org/wiki/GEOPACKAGEswg/06-May-2019  - it's quite possible somebody else was there or came in late and was not logged as present - this is merely an observation.

Bottom line, issue#1 removal of AUTOINCREMENT breaks existing software.

**Comment by jyutzler** [comment]
After a thorough investigation, I do not believe any action is required here.

1. Your understanding of AUTOINCREMENT is incorrect. Most of the time, [you do not need it](https://sqlite.org/autoinc.html). You can try this on your own - create a table without AUTOINCREMENT, add a row with a null primary key, delete the row, and add another. SQLite will manage the keys for you and assign the primary key value `1` to the second row (1 greater than the highest key already in use). The only time you need AUTOINCREMENT is if you are doing a lot of inserting and deleting and you want to prevent a primary key value from being reused.

2. I don't think believe we need to take any action on AUTOINCREMENT. It does not appear in any of the normative table definitions in Annex C, just the informative ones. Even in v1.0.1, the [DDL for a features table](http://www.geopackage.org/spec101/index.html#_sample_feature_table_informative) was always informative. The ATS (http://www.geopackage.org/spec121/index.html#_vector_features_user_data_tables et. al.) and [ETS](https://github.com/opengeospatial/ets-gpkg12/blob/master/src/main/java/org/opengis/cite/gpkg12/CommonFixture.java#L199) don't test for it.

3. It is unclear to me how [Requirement 150](http://www.geopackage.org/spec/index.html#r150) would violate the modular specification. Technically you can have whatever type you want as a primary key and it can be in any position in the table. We are just adding another path that would allow views to function. Supporting views in GeoPackage has been part of the intent from the beginning (the words "or view" appear all over the 1.0.1 version) and multiple OGC members have directly asked me how to do it. Since primary keys don't exist for views in SQLite, it wasn't possible to implement a view with a strict interpretation of the requirements of v1.0, 1.1, or 1.2. PR 449 was designed to fix this as best as we could. Of course if you try to update a view it won't work, but IMHO that is working as designed. 

4. This was a clerical error - I just failed to put a trailing `*` after Kevin and Tracey's names to indicate their attendance. Tracey seconded a motion and that is proof she was in fact on the call. I have corrected the meeting notes.

**Comment by jerstlouis** [comment]
From https://www.sqlite.org/autoinc.html :

> the purpose of AUTOINCREMENT is to prevent the reuse of ROWIDs from previously deleted rows.

**Comment by jyutzler** [comment]
In today's SWG meeting, @KRyden requested more time to review the issue.

**Comment by jyutzler** [comment]
#517 should not be closed until this is completely adjudicated

**Comment by fjlopez** [comment]
Regarding to this quirk:

> Is just a SQLite quirk I noticed, and not something I thought was a serious issue. But if you don't declare an INTEGER PRIMARY KEY column as NOT NULL, it behaves just the same, but when you ask SQLite to describe the column, it doesn't report it as NOT NULL:

It is not a quirk, it is just bad documentation in the SQLite website. The description of the pragma `table_info` at https://www.sqlite.org/pragma.html#pragma_table_info is deceitful. 

> This pragma returns one row for each column in the named table. Columns in the result set include the column name, data type, **whether or not the column can be NULL**, and the default value for the column. 

In the source code is more clear: 

> notnull:    True if 'NOT NULL' is part of column declaration

https://github.com/sqlite/sqlite/blob/b5aaee5e31c081b3ba5beeb563419f2d66be2656/src/pragma.c#L1135

The behaviour is as expected, the documentation is not.

**Comment by PeterAronson** [comment]
Thanks -- I'm actually keeping an eye on the page, but I might not have noticed this for a while. I see my post has been mostly favorably received -- even this last guy seems more  annoyed at SQLite than me...

Peter
________________________________

**Comment by jyutzler** [comment]
Regarding 2, I put language into https://github.com/opengeospatial/geopackage/pull/534/commits/1a28a78881f3899a68aee4eb0e49160a304bd5a0 indicating that the rowid alias SHALL be present. I'm not sure how to test this for this, though.

On 3, we run into a testability issue. If we test for notnull, we would mark GeoPackages that don't set that flag as invalid even though there is no damage done, no risk to interoperability.

**Comment by PeterAronson** [comment]
(Oops! Sorry about the reply.)

@fjlopez 
Actually, I was considering the *behavior* to be a bit of a quirk, but thinking about it, you may be correct. Here's a survey of SQLite's behavior in these cases:
```SQL
create table pk_is_internal_rowid_alias (id integer primary key);
pragma table_info (pk_is_internal_rowid_alias);
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          integer     0                       1
insert into pk_is_internal_rowid_alias values (NULL);
select * from pk_is_internal_rowid_alias;
id
----------
1

create table pk_is_internal_rowid_alias_nn (id integer not null primary key);
pragma table_info (pk_is_internal_rowid_alias_nn);
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          integer     1                       1
insert into pk_is_internal_rowid_alias_nn values (NULL);
select * from pk_is_internal_rowid_alias_nn;
id
----------
1

create table pk_is_not_internal_rowid_alias (id integer primary key desc);
pragma table_info (pk_is_not_internal_rowid_alias);
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          integer     0                       1
insert into pk_is_not_internal_rowid_alias values (NULL);
select * from pk_is_not_internal_rowid_alias;
id
----------


create table pk_is_not_internal_rowid_alias_nn (id integer not null primary key desc);
pragma table_info (pk_is_not_internal_rowid_alias_nn);
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          integer     1                       1
insert into pk_is_not_internal_rowid_alias_nn values (NULL);
Error: near line 18: NOT NULL constraint failed: pk_is_not_internal_rowid_alias_nn.id

create table pk_is_not_internal_rowid_alias_wr (id integer primary key) without rowid;
pragma table_info (pk_is_not_internal_rowid_alias_wr);
cid         name        type        notnull     dflt_value  pk
----------  ----------  ----------  ----------  ----------  ----------
0           id          integer     1                       1
insert into pk_is_not_internal_rowid_alias_wr values (NULL);
Error: near line 22: NOT NULL constraint failed: pk_is_not_internal_rowid_alias_wr.id
sqlite>
```
It seems a bit inconsistent in places.

**Comment by PeterAronson** [comment]
@jyutzler This section of the SQLite documentation gives you some background: https://www.sqlite.org/lang_createtable.html#rowid.

I think the following criteria should work:
1. The table declaration in the `sql` column of the `sqlite_master` table does not end with `without rowid` (case-insensitive).
2. The id column id is the sole member of the primary key as indicated by `table_info` pragma.
3. The id column is declared `integer`.
4. The string `primary key desc` (case insensitive) does not occur in the `sql` column of `sqlite_master` row for the table.

Both 3. and 4. could be replaced by requiring the string `<id_column> integer primary key` (case-insensitive), where `<id_column>` is replaced by the actual name of the id column, be present in the `sql` column of `sqlite_master` entry for the table declaration.

**Comment by PeterAronson** [comment]
@jyutzler BTW, one possible minor issue with dealing with the alternate approach above, is SQLite allows 4 different ways of delimiting identifiers (5 if you count no delimiter):
```SQL
sqlite> select sql from sqlite_master;
sql
-----------------------------------------------------------
CREATE TABLE t1 (id integer primary key)
CREATE TABLE t2 ("id" integer primary key)
CREATE TABLE t3 ([id] integer primary key)
CREATE TABLE t4 (`id` integer primary key)
CREATE TABLE t5 ('id' integer primary key)
```
So you would have to test for any of those.

**Comment by fjlopez** [comment]
I think that a query for the rowid column can also help for the `without rowid`.  
```sql
sqlite> create table id_pk(id integer primary key);
sqlite> select rowid from id_pk;
sqlite> create table id_pk_without_rowid(id integer primary key) without rowid;
sqlite> select rowid from id_pk_without_rowid;
Error: no such column: rowid
```

**Comment by fjlopez** [comment]
`pragma index_list(table)` can reveal a `desc` for the primary column in such table. This can be an alternative to 4.

**Comment by PeterAronson** [comment]
@fjlopez Unless, of course, they create a column named `rowid`.
```SQL
sqlite> create table gotcha (rowid integer primary key) without rowid;
sqlite> select rowid from gotcha;
sqlite>
```
However, using `_rowid_` is probably safe enough -- who would name a column *that* (unless they were actively trying to be malicious). 

The pragma approach looks like a good alternative. If I were writing from a C/C++ that is what I would probably use.

**Comment by fjlopez** [comment]
The pragma approach also works in the Xerial JDBC driver and since SQLITE 3.16 it is possible to use pragma functions:

```sql
sqlite> select * from pragma_table_info('id_pk');
0|id|integer|0||1
```

**Comment by jyutzler** [comment]
In light of https://github.com/opengeospatial/geopackage/issues/446#issuecomment-655071820, is there any further action required in this ticket?

**Comment by PeterAronson** [comment]
@jyutzler I do not believe there is any further action required for this ticket. The Pull Request you referenced handles the issues very cleanly. Thanks you!
